compiler/ballerina-lang/src/main/java/io/ballerina/projects/CodeGenOptimizationReportEmitter.java

```diff
@@ -0,0 +1,177 @@
+ *
+ *  WSO2 LLC. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.ballerina.projects;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import org.ballerinalang.model.elements.PackageID;
+import org.ballerinalang.model.symbols.SymbolOrigin;
+import org.wso2.ballerinalang.compiler.bir.model.BIRNode;
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.PrintStream;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Emits time durations and optimized node details for codegen optimization.
+ *
+ * @since 2201.10.0
+ */
+public final class CodeGenOptimizationReportEmitter {
+
+    private static final CompilerContext.Key<CodeGenOptimizationReportEmitter> CODEGEN_OPTIMIZATION_REPORT_EMITTER_KEY =
+            new CompilerContext.Key<>();
+    private final Map<PackageID, Long> birOptimizationDurations;
+    private long nativeOptimizationDuration;
+    private final PrintStream out;
+    private static final String CODEGEN_OPTIMIZATION_REPORT = "codegen_optimization_report.json";
+
+    private CodeGenOptimizationReportEmitter(CompilerContext compilerContext) {
+        compilerContext.put(CODEGEN_OPTIMIZATION_REPORT_EMITTER_KEY, this);
+        this.out = System.out;
+        this.birOptimizationDurations = new HashMap<>();
+        this.nativeOptimizationDuration = 0;
+    }
+
+    public static CodeGenOptimizationReportEmitter getInstance(CompilerContext compilerContext) {
+        CodeGenOptimizationReportEmitter codegenOptimizationReportEmitter =
+                compilerContext.get(CODEGEN_OPTIMIZATION_REPORT_EMITTER_KEY);
+        if (codegenOptimizationReportEmitter == null) {
+            codegenOptimizationReportEmitter = new CodeGenOptimizationReportEmitter(compilerContext);
+        }
+        return codegenOptimizationReportEmitter;
+    }
+
+    protected void flipBirOptimizationTimer(PackageID pkgId) {
+        if (this.birOptimizationDurations.containsKey(pkgId)) {
+            long totalDuration = System.currentTimeMillis() - this.birOptimizationDurations.get(pkgId);
```

Comment:
Let move `System.currentTimeMillis` out of the map lookup

Reply:
Fixed in eb164e9b0bfd3dac61e1b00a86d30e96f11b33a2.